test/unit/utils/cuda/cuda_ptr_test.cpp

```diff
@@ -0,0 +1,222 @@
+    CHECK_CUDA_DRIVER_ERROR(cuMemCreate(&handle, padded_size, &prop, 0),
+                            "Failed to create allocation");
+
+    // Reserve the memory address
+    CHECK_CUDA_DRIVER_ERROR(cuMemAddressReserve(&addr, padded_size,
+                                                granularity, 0, 0),
+                            "Failed to reserve address");
+
+    // Map the memory
+    CHECK_CUDA_DRIVER_ERROR(cuMemMap(addr, padded_size, 0, handle, 0),
+                            "Failed to map memory");
+
+    cout << "Address: " << std::hex << std::showbase << addr
+              << " Buffer size: " << std::dec << len
+              << " Padded size: " << std::dec << padded_size << std::endl;
+
+    // // Set the memory access rights
+    // CUmemAccessDesc access = {};
+    // access.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+    // access.location.id = dev_id;
+    // access.flags = CU_MEM_ACCESS_FLAGS_PROT_READWRITE;
+    // CHECK_CUDA_DRIVER_ERROR(cuMemSetAccess(addr, len, &access, 1),
+    //                         "Failed to set access");
+
+    _addr = (void*)addr;
+}
+
+void releaseVMM(int dev_id, size_t len, void* addr)
+{
+    CHECK_CUDA_ERROR(cudaSetDevice(dev_id), "Failed to set device");
+    CHECK_CUDA_DRIVER_ERROR(cuMemUnmap((CUdeviceptr)addr, padded_size),
+                            "Failed to unmap memory");
+    CHECK_CUDA_DRIVER_ERROR(cuMemRelease(handle),
+                            "Failed to release memory");
+    CHECK_CUDA_DRIVER_ERROR(cuMemAddressFree((CUdeviceptr)addr, padded_size),
+                            "Failed to free reserved address");
+}
+
+#endif
+
+
+int main()
+{
+    void *address;
+    size_t len = 1024;
+
+    /* Discover environment */
+    int ngpus;
+    cudaGetDeviceCount(&ngpus);
+
+    if (!ngpus) {
+        cout << "No GPGPU devices detected, nothing to test!" << endl;
+        return 0;
+    }
+
+    /* Test regular CUDA malloc */
+    {
+        cout << endl << "*************************" << endl;
+        cout << "      Test malloc'd memory" << endl;
+
+        address = malloc(len);
+        assert(address);
+        std::unique_ptr<nixlCudaPtrCtx> ctx =
+                nixlCudaPtrCtx::nixlCudaPtrCtxInit(address);
+        assert(ctx->getMemType() == nixlCudaPtrCtx::MEM_HOST);
+        cout << " >>>> PASSED! <<<<<<<" << endl;
+        free(address);
+        cout << "*************************" << endl;
+    }
+
+    /* Test regular CUDA malloc */
+    {
+        cout << endl << "*************************" << endl;
+        cout << "      Test CUDA malloc'd memory" << endl;
+
+        allocateCUDA(0, len, address);
+        std::unique_ptr<nixlCudaPtrCtx> ctx =
+                nixlCudaPtrCtx::nixlCudaPtrCtxInit(address);
+        assert(ctx->getMemType() == nixlCudaPtrCtx::MEM_DEV);
+        cout << " >>>> PASSED! <<<<<<<" << endl;
+        releaseCUDA(0, address);
+        cout << "*************************" << endl;
+    }
+
+#ifdef HAVE_CUDA_VMM
```

Comment:
should we use HAVE_CUDA_VMM also in cuda_utils.cpp?

Reply:
There we have a more fine-grained set of macros that I derived from UCX

Reply:
I'm not familiar with gtest.
If possible, I'd prefer to do it as a separate PR. But I'll do it.
Right now, CI won't run it anyway as it requires CUDA devices.